### PR TITLE
Ispn 5195 ispn 5205

### PR DESCRIPTION
**Pull request by slaskawi**
Hey!

This is a rebase of fix provided by @marciopd [1]. It solved the following problems:
* https://issues.jboss.org/browse/ISPN-5205
* https://issues.jboss.org/browse/ISPN-5195

Thanks @marciopd for providing the fix and the reproducer!

[1] https://github.com/infinispan/infinispan/pull/3292

### CODE REVIEW

**Comment by slaskawi** [comment]
Hey @galderz !

Since the JCache codebase changed significantly, could you please take a look if I ported it correctly?

Thanks!

**Comment by marciopd** [comment]
You're welcome, @slaskawi !

**Comment by galderz** [comment]
@slaskawi Checking...

File: jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java [hunk 1]
```diff
@@ -1,24 +1,28 @@
 package org.infinispan.jcache.annotation;
 
-import java.lang.annotation.Annotation;
```
Comment (galderz) [inline jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java]: Can we please stop reordering imports??

File: jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java [hunk 2]
```diff
@@ -97,13 +120,34 @@ private JCacheManager toJCacheManager(EmbeddedCacheManager cacheManager) {
...
-      if (cache != null)
-         return cache;
+   private <K, V> Cache<K, V> getCacheFromDefaultCacheManager(final String cacheName) {
+      final Configuration defaultInjectedConfiguration = getBeanReference(Configuration.class);
+
+      getDefaultCacheManager().defineConfiguration(cacheName,
+            defaultInjectedConfiguration);
+
+      return getDefaultJCacheManager().getOrCreateCache(cacheName,
+                                                      getDefaultCacheManager().<K, V>getCache(cacheName).getAdvancedCache());
```
Comment (galderz) [inline jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java]: Formatting...

File: jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java [hunk 3]
```diff
@@ -35,36 +39,55 @@
...
    private JCacheManager defaultJCacheManager;
 
+   private EmbeddedCacheManager getDefaultCacheManager() {
+      if (defaultCacheManager == null) {
+         defaultCacheManager = getBeanReference(EmbeddedCacheManager.class);
+      }
+      return defaultCacheManager;
+   }
+
+   private JCacheManager getDefaultJCacheManager() {
```
Comment (galderz) [inline jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java]: This method is unnecessary, just reference defaultCacheManager, which is private. Same with setter.

File: jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java [hunk 4]
```diff
@@ -97,13 +120,34 @@ private JCacheManager toJCacheManager(EmbeddedCacheManager cacheManager) {
       return getCacheFromDefaultCacheManager(cacheName);
    }
 
-   private <K, V> Cache<K, V> getCacheFromDefaultCacheManager(String cacheName) {
-      Cache<K, V> cache = defaultJCacheManager.getCache(cacheName);
-      if (cache != null)
-         return cache;
+   private <K, V> Cache<K, V> getCacheFromDefaultCacheManager(final String cacheName) {
+      final Configuration defaultInjectedConfiguration = getBeanReference(Configuration.class);
```
Comment (galderz) [inline jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java]: Hmmm, do we support injecting Configuration instances?

Comment (marciopd) [inline jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java]: Hi @galderz , 
Infinispan produces a default configuration, which can be overriden[1].
 
[1] [19.2.2. Override the default embedded cache manager and configuration](http://infinispan.org/docs/7.0.x/user_guide/user_guide.html#_override_the_default_embedded_cache_manager_and_configuration)

Comment (galderz) [inline jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java]: Ah yes, producer! Thanks for reminder :)

**Comment by galderz** [comment]
Thanks @marciopd, all minor issues left now. Can you sort out those small issues with formatting and put a comment when it's done?

**Comment by galderz** [comment]
Testing looks good

File: jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java (hunk 3 above)
Comment (vchepeli) [inline jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java]: Agree with @galderz . Adding setters/getters does not make sense here

File: jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java (hunk 4 above)
Comment (vchepeli) [inline jcache/embedded/src/main/java/org/infinispan/jcache/annotation/InjectedCacheResolver.java]: yep in cdi tests we override default producer and use custom one wich produces configurations for caches

**Comment by vchepeli** [comment]
Looks good for me

**Comment by marciopd** [comment]
Hi @slaskawi,

I checked out your pull request branch and did all the changes suggested by the team.

I can't commit on this branch, so I'll just post the code below, ok?
If there's a better way of sharing, I'll be happy to help.


```
package org.infinispan.jcache.annotation;

import java.lang.annotation.Annotation;
import java.net.URI;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

import javax.cache.Cache;
import javax.cache.Caching;
import javax.cache.annotation.CacheInvocationContext;
import javax.cache.annotation.CacheResolver;
import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.context.spi.CreationalContext;
import javax.enterprise.inject.spi.Bean;
import javax.enterprise.inject.spi.BeanManager;
import javax.inject.Inject;

import org.infinispan.cdi.InfinispanExtension;
import org.infinispan.cdi.InfinispanExtensionEmbedded;
import org.infinispan.cdi.util.BeanManagerProvider;
import org.infinispan.configuration.cache.Configuration;
import org.infinispan.configuration.global.GlobalConfiguration;
import org.infinispan.jcache.embedded.JCacheManager;
import org.infinispan.manager.EmbeddedCacheManager;

/**
 * Injected cache resolver for situations where caches and/or cache managers
 * are injected into the CDI beans. In these situations, bridging is required
 * in order to bridge between the Infinispan based caches and the JCache
 * cache instances which is what it's expected by the specification cache
 * resolver.
 *
 * @author Galder Zamarreño
 * @since 5.3
 */
@ApplicationScoped
public class InjectedCacheResolver implements CacheResolver {

   private EmbeddedCacheManager defaultCacheManager;

   private final Map<EmbeddedCacheManager, JCacheManager> jcacheManagers = new HashMap<>();
   private JCacheManager defaultJCacheManager;

   // for proxy.
   public InjectedCacheResolver() {
   }

   @Inject
   public InjectedCacheResolver(final InfinispanExtension extension, final BeanManager beanManager) {
      final Set<InfinispanExtensionEmbedded.InstalledCacheManager> installedCacheManagers = extension.getEmbeddedExtension().getInstalledEmbeddedCacheManagers(beanManager);
      for (final InfinispanExtensionEmbedded.InstalledCacheManager installedCacheManager : installedCacheManagers) {
         final JCacheManager jcacheManager = toJCacheManager(installedCacheManager.getCacheManager());
         this.jcacheManagers.put(installedCacheManager.getCacheManager(), jcacheManager);
      }
      initializeDefaultCacheManagers();
   }

   private void initializeDefaultCacheManagers() {
      defaultCacheManager = getBeanReference(EmbeddedCacheManager.class);

      if (jcacheManagers.containsKey(defaultCacheManager)) {
         defaultJCacheManager = jcacheManagers.get(defaultCacheManager);
      } else {
         defaultJCacheManager = toJCacheManager(defaultCacheManager);
         jcacheManagers.put(defaultCacheManager, defaultJCacheManager);
      }
   }

   private JCacheManager toJCacheManager(final EmbeddedCacheManager cacheManager) {
      final GlobalConfiguration globalCfg = cacheManager.getCacheManagerConfiguration();
      final String name = globalCfg.globalJmxStatistics().cacheManagerName();
      return new JCacheManager(URI.create(name), cacheManager, Caching.getCachingProvider());
   }

   @Override
   public <K, V> Cache<K, V> resolveCache(final CacheInvocationContext<? extends Annotation> cacheInvocationContext) {
      Contracts.assertNotNull(cacheInvocationContext, "cacheInvocationContext parameter must not be null");

      final String cacheName = cacheInvocationContext.getCacheName();

      // If the cache name is empty the default cache of the default cache manager is returned.
      if (cacheName.trim().isEmpty()) {
         return getCacheFromDefaultCacheManager(cacheName);
      }

      // Iterate on all cache managers because the cache used by the
      // interceptor could use a specific cache manager.
      for (final EmbeddedCacheManager cm : jcacheManagers.keySet()) {
         final Set<String> cacheNames = cm.getCacheNames();
         for (final String name : cacheNames) {
            if (name.equals(cacheName)) {
               final JCacheManager jcacheManager = jcacheManagers.get(cm);
               final Cache<K, V> cache = jcacheManager.getCache(cacheName);
               if (cache != null)
                  return cache;

               return jcacheManager.getOrCreateCache(
                     cacheName, cm.<K, V>getCache(cacheName).getAdvancedCache());
            }
         }
      }

      // If the cache has not been defined in the default cache manager
      // or in a specific one a new cache is created in the default
      // cache manager with the default configuration.
      return getCacheFromDefaultCacheManager(cacheName);
   }

   private <K, V> Cache<K, V> getCacheFromDefaultCacheManager(final String cacheName) {
      final Configuration defaultInjectedConfiguration = getBeanReference(Configuration.class);

      defaultCacheManager.defineConfiguration(cacheName, defaultInjectedConfiguration);

      return defaultJCacheManager.getOrCreateCache(cacheName, defaultCacheManager.<K, V> getCache(cacheName)
            .getAdvancedCache());
   }

   private BeanManager getBeanManager() {
      return BeanManagerProvider.getInstance().getBeanManager();
   }

   @SuppressWarnings("unchecked")
   private <T> T getBeanReference(final Class<T> beanType) {
      final BeanManager bm = getBeanManager();
      final Iterator<Bean<?>> iterator = bm.getBeans(beanType).iterator();
      if (!iterator.hasNext()) {
         throw new IllegalStateException(String.format(
               "Default bean of type %s not found.", beanType.getName()));
      }

      final Bean<?> configurationBean = iterator.next();
      final CreationalContext<?> createCreationalContext = bm.createCreationalContext(configurationBean);
      return (T) bm.getReference(configurationBean, beanType, createCreationalContext);
   }

}
```

**Comment by slaskawi** [comment]
Unfortunately those 2 commits have been already merged, so I created another PR: https://github.com/infinispan/infinispan/pull/3349

Thanks @marciopd ! Closing this one.

**Pull request closed without merging**